tests: Port `symbol-mangling-hashed` to rmake.rs

jieyouxu opened the pull request:
Part of #121876.

This PR supersedes #128567 and is co-authored with @lolbinarycat.

### Summary

This PR ports `tests/run-make/symbol-mangling-hashed` to rmake.rs. Notable differences when compared to the Makefile version includes:

- It's no longer limited to linux + x86_64 only. In particular, this now is exercised on darwin and windows (esp. msvc) too.
- The test uses `object` crate to be more precise in the filtering, and avoids relying on parsing the human-readable `nm` output for *some* `nm` in the given environment (which isn't really a thing on msvc anyway, and `llvm-nm` doesn't handle msvc dylibs AFAICT).
- Dump the symbols satisfying various criteria on test failure to make it hopefully less of a pain to debug if it ever fails in CI.

### Review advice

- Best reviewed commit-by-commit.
- I'm not *super* sure about the msvc logic, would benefit from a MSVC (PE/COFF) expert taking a look.

---

try-job: x86_64-msvc-1
try-job: i686-msvc-1
try-job: i686-mingw
try-job: x86_64-mingw-1
try-job: x86_64-apple-1
try-job: aarch64-apple
try-job: test-various

jieyouxu committed:
run-make-support: add `-Csymbol-mangling-version` and `-Cprefer-dynamic` helpers to rustc

Co-authored-by: binarycat <binarycat@envs.net>

jieyouxu committed:
run-make-support: collapse re-export

jieyouxu committed:
run-make-support: add `is_windows_gnu` helper

jieyouxu committed:
run-make-support: add `object`-based symbol helpers

- `dynamic_symbol_names`
- `text_section_global_dynamic_symbol_names`
- `global_undefined_dynamic_symbol_names`

Also add some missing `#[track_caller]` attributes.

Co-authored-by: binarycat <binarycat@envs.net>

jieyouxu committed:
tests: port `symbol-mangling-hashed` to rmake.rs

- Use `object` based test logic instead of processing `nm`
  human-readable textual output.
- Try to expand test coverage to not be limited to only linux + x86_64.

Co-authored-by: binarycat <binarycat@envs.net>

jieyouxu commented:
(Sorry Mark, this isn't ready for review yet)

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 10841e5672d450eeb5d605fca7de5255452bc767 with merge cd8beabdd1dd7ddfbca8933f0eaec87cca0f9a57...
<!-- homu: {"type":"TryBuildStarted","head_sha":"10841e5672d450eeb5d605fca7de5255452bc767","merge_sha":"cd8beabdd1dd7ddfbca8933f0eaec87cca0f9a57"} -->

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 10841e5672d450eeb5d605fca7de5255452bc767 with merge a5e58b62e33adcba8b55845200299fa5d121108e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"10841e5672d450eeb5d605fca7de5255452bc767","merge_sha":"a5e58b62e33adcba8b55845200299fa5d121108e"} -->

jieyouxu commented on tests/run-make/symbol-mangling-hashed/rmake.rs:
```diff
@@ -0,0 +1,264 @@
+// ignore-tidy-linelength
+//! Basic smoke test for the unstable option `-C symbol_mangling_version=hashed` which aims to
+//! replace full symbol mangling names based on hash digests to shorten symbol name lengths in
+//! dylibs for space savings.
```
Improvement suggestions for this test comment are welcomed. The original Makefile had zero comments so this is already more context than no comments...

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12869380616/job/35878072780) [(plain)](https://github.com/rust-lang-ci/rust/commit/a5e58b62e33adcba8b55845200299fa5d121108e/checks/35878072780/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (50525/51556)
Updating files:  99% (51041/51556)
Updating files: 100% (51556/51556)
Updating files: 100% (51556/51556), done.
Switched to a new branch 'try'
branch 'try' set up to track 'origin/try'.
[command]/opt/homebrew/bin/git log -1 --format=%H
a5e58b62e33adcba8b55845200299fa5d121108e
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
---- [run-make] tests/run-make/symbol-mangling-hashed stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/symbol-mangling-hashed/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage1-tools-bin/cargo" CC="/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx11 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk" CXX="/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx11 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -stdlib=libc++" DYLD_LIBRARY_PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0-bootstrap-tools/aarch64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0/lib:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2-tools-bin:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib" HOST_RPATH_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib" LD_LIB_PATH_ENVVAR="DYLD_LIBRARY_PATH" LLVM_BIN_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegendata codegentypes core coroutines coverage csky cskyasmparser cskycodegen cskydesc cskydisassembler cskyinfo debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin/FileCheck" NODE="/opt/homebrew/bin/node" PYTHON="/usr/bin/python3" RUSTC="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustc" RUSTDOC="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustdoc" SOURCE_ROOT="/Users/runner/work/rust/rust" TARGET="aarch64-apple-darwin" TARGET_RPATH_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib" TARGET_RPATH_ENV="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/symbol-mangling-hashed/rmake_out:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0-bootstrap-tools/aarch64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0/lib" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/symbol-mangling-hashed/rmake"
--- stdout -------------------------------
checking dylib `liba_dylib.dylib`
--- stderr -------------------------------
global dynamic symbols in `.text` section: []

thread 'main' panicked at /Users/runner/work/rust/rust/tests/run-make/symbol-mangling-hashed/rmake.rs:96:17:
thread 'main' panicked at /Users/runner/work/rust/rust/tests/run-make/symbol-mangling-hashed/rmake.rs:96:17:
expected two global dynamic symbols starting with `_RNxC7a_dylib` in `.text` section
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12869380616/job/35882554742)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12869380616/job/35882554742","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 897a14517daf10f2b44e139457e534ed55f5d3a2 with merge 06b585477f7bb29f168aad77790aa344885ca087...
<!-- homu: {"type":"TryBuildStarted","head_sha":"897a14517daf10f2b44e139457e534ed55f5d3a2","merge_sha":"06b585477f7bb29f168aad77790aa344885ca087"} -->

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12871602860/job/35885260919) [(plain)](https://github.com/rust-lang-ci/rust/commit/06b585477f7bb29f168aad77790aa344885ca087/checks/35885260919/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (50525/51556)
Updating files:  99% (51041/51556)
Updating files: 100% (51556/51556)
Updating files: 100% (51556/51556), done.
Switched to a new branch 'try'
branch 'try' set up to track 'origin/try'.
[command]/opt/homebrew/bin/git log -1 --format=%H
06b585477f7bb29f168aad77790aa344885ca087
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
---- [run-make] tests/run-make/symbol-mangling-hashed stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/symbol-mangling-hashed/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage1-tools-bin/cargo" CC="/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx11 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk" CXX="/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx11 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -stdlib=libc++" DYLD_LIBRARY_PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0-bootstrap-tools/aarch64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0/lib:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2-tools-bin:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib" HOST_RPATH_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib" LD_LIB_PATH_ENVVAR="DYLD_LIBRARY_PATH" LLVM_BIN_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegendata codegentypes core coroutines coverage csky cskyasmparser cskycodegen cskydesc cskydisassembler cskyinfo debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin/FileCheck" NODE="/opt/homebrew/bin/node" PYTHON="/usr/bin/python3" RUSTC="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustc" RUSTDOC="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustdoc" SOURCE_ROOT="/Users/runner/work/rust/rust" TARGET="aarch64-apple-darwin" TARGET_RPATH_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib" TARGET_RPATH_ENV="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/symbol-mangling-hashed/rmake_out:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0-bootstrap-tools/aarch64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0/lib" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/symbol-mangling-hashed/rmake"
--- stdout -------------------------------
checking dylib `liba_dylib.dylib`
--- stderr -------------------------------
exported dynamic symbols: [
exported dynamic symbols: [
    "__RNxC7a_dylib12Hei0S60eWdBH",
    "__RNxC7a_dylib12HjermeVgSqiY",
    "_rust_metadata_a_dylib_43019022ffda618d",

thread 'main' panicked at /Users/runner/work/rust/rust/tests/run-make/symbol-mangling-hashed/rmake.rs:59:13:
thread 'main' panicked at /Users/runner/work/rust/rust/tests/run-make/symbol-mangling-hashed/rmake.rs:59:13:
expected two dynamic symbols starting with `_RNxC7a_dylib`
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12871602860/job/35888955293)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12871602860/job/35888955293","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit edf0248ff4273a3f71acae5afe6e1af0af17efe2 with merge 325a92d5d2ce268f1bd74c77d1319ea96a8947ce...
<!-- homu: {"type":"TryBuildStarted","head_sha":"edf0248ff4273a3f71acae5afe6e1af0af17efe2","merge_sha":"325a92d5d2ce268f1bd74c77d1319ea96a8947ce"} -->

Noratrieb commented on tests/run-make/symbol-mangling-hashed/rmake.rs:
```diff
@@ -0,0 +1,123 @@
+
+fn main() {
+    rustc()
+        .input("a_dylib.rs")
```
instead of a and b, maybe call them default and hashed?

jieyouxu replied:
Good point, renamed `a` -> `hashed` and `b` -> `default`.

Noratrieb commented on tests/run-make/symbol-mangling-hashed/rmake.rs:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    {
+        let so_filename = dynamic_lib_name("b_dylib");
```
you called it dylib above and so here

jieyouxu replied:
Changed to dylib for consistency.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12880830997/job/35915795742)
Build commit: 325a92d5d2ce268f1bd74c77d1319ea96a8947ce (`325a92d5d2ce268f1bd74c77d1319ea96a8947ce`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12880830997/job/35915795742"},"merge_sha":"325a92d5d2ce268f1bd74c77d1319ea96a8947ce"} -->

jieyouxu commented:
Since you're already looking at it... r? @Noratrieb (feel free to reroll)

jieyouxu commented on src/tools/run-make-support/src/lib.rs:
The formatting in this `lib.rs` is nasty, intended to cleanup in a follow-up.

jieyouxu commented:
Renamed the files to `hashed` and `default` instead of `a` and `b`, and addressed the `so` -> `dylib` naming inconsistency.

Noratrieb reviewed (changes requested):
The new test doesn't exactly test what the previous test did, but also I'm a bit unsure of what exactly we should test in the first place.
To me the absence of the hello string should be good enough? But maybe not?

jieyouxu commented on tests/run-make/symbol-mangling-hashed/Makefile:
```diff
@@ -1,48 +0,0 @@
-	[ "$$($(NM) $(TMPDIR)/$(SO_NAME) | grep _RNxC6a_rlib | grep -c ' T ')" -eq "2" ]
-	[ "$$($(NM) $(TMPDIR)/$(SO_NAME) | grep _RNxC7a_dylib | grep -c ' U ')" -eq "1" ]
-
-	[ "$$($(NM) $(TMPDIR)/$(BIN_NAME) | grep _RNxC6a_rlib | grep -c ' U ')" -eq "1" ]
```
```makefile
[ "$$($(NM) $(TMPDIR)/$(SO_NAME) | grep _RNxC6a_rlib | grep -c ' T ')" -eq "2" ]
```

is

```rs
        let expected_rlib_prefix =
            if is_darwin() { "__RNxC11hashed_rlib" } else { "_RNxC11hashed_rlib" };
        if dynamic_symbols.iter().filter(|sym| sym.starts_with(expected_rlib_prefix)).count() != 2 {
            eprintln!("exported dynamic symbols: {:#?}", dynamic_symbols);
            panic!("expected two exported symbols starting with `{expected_rlib_prefix}`");
        }
```

Note that renaming `a_rlib` -> `hashed_rlib` changes the mangled symbol name from `_RNxC6a_rlib` to `__RNxC11hashed_rlib`.

The other *counts* were converted to existence checks which aren't equivalent, I'll fix that.

jieyouxu commented:
Changed the test to actually run the binary (I completely missed that line, oops). Did not change the globally-undefined symbol checks, unsure what they actually *want* to check.
@rustbot review

jieyouxu commented:
Actually let me run a try-job first.
@bors try
@rustbot author

bors commented:
:hourglass: Trying commit 2aa5888a401be70d2e50259812a0dae43f7d6397 with merge 21093ea86886cf8c671431b635648f1de195b21a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2aa5888a401be70d2e50259812a0dae43f7d6397","merge_sha":"21093ea86886cf8c671431b635648f1de195b21a"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13011880883/job/36291306049) [(plain)](https://github.com/rust-lang-ci/rust/commit/21093ea86886cf8c671431b635648f1de195b21a/checks/36291306049/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
x86_64-apple-1
aarch64-apple
test-various
##[endgroup]
INFO:root:Job type: TryRunType(custom_jobs=['x86_64-msvc', 'i686-msvc-1', 'i686-mingw', 'x86_64-mingw-1', 'x86_64-apple-1', 'aarch64-apple', 'test-various'])
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 314, in <module>
    calculate_job_matrix(data)
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 266, in calculate_job_matrix
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 266, in calculate_job_matrix
    jobs = calculate_jobs(run_type, job_data)
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 153, in calculate_jobs
    raise Exception(
    raise Exception(
Exception: Custom job(s) `['x86_64-msvc']` not found in auto jobs
##[error]Process completed with exit code 1.
```

</details>

bors commented:
:hourglass: Trying commit 2aa5888a401be70d2e50259812a0dae43f7d6397 with merge 840317a990493de24cf139e520ebc5459b787864...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2aa5888a401be70d2e50259812a0dae43f7d6397","merge_sha":"840317a990493de24cf139e520ebc5459b787864"} -->

Noratrieb commented on src/tools/run-make-support/src/symbols.rs:
```diff
@@ -2,18 +2,32 @@ use std::path::Path;
-/// Uses a callback because `SymbolIterator` does not own its data.
+/// Given an [`object::File`], find the exported dynamic symbol names via
+/// [`object::Object::exports`]. This does **not** impose any filters on the specific dynamic
+/// symbols, e.g. if they are global or local, if they are defined or not, and in which section the
```
I don't know about other formats but at least on ELF `exports()` appears to only return defined symbols

jieyouxu replied:
Fixed comment

Noratrieb commented on tests/run-make/symbol-mangling-hashed/rmake.rs:
```diff
@@ -0,0 +1,130 @@
+            if is_darwin() { "__RNxC11hashed_rlib" } else { "_RNxC11hashed_rlib" };
+        let expected_dylib_prefix =
+            if is_darwin() { "__RNxC12hashed_dylib" } else { "_RNxC12hashed_dylib" };
+        if dynamic_symbols.iter().any(|sym| {
```
this just checks that a binary doesn't export any symbols, which isn't a very useful thing to test for this test.
I think we should just drop this binary part.

jieyouxu replied:
Dropped symbol checks on the bin

Noratrieb commented on tests/run-make/symbol-mangling-hashed/rmake.rs:
```diff
@@ -0,0 +1,130 @@
+
+        let dynamic_symbols = exported_dynamic_symbol_names(&dylib_file);
+
+        if dynamic_symbols.iter().filter(|sym| sym.contains("hello")).count() != 0 {
```
I would like to have different variants of hello for the different crates to ensure that searching for one won't find another.

jieyouxu replied:
Changed the `hello` fn name of the various dylib/rlibs to be called `{hd,hr,dd}hello`, and adjusted the checks.

Noratrieb reviewed (changes requested):
Okay, I think that's it. I took a look at some stuff and I think I've figured it out now.
We should drop the last part of the test, fix the comment on the symbols helper and use different identifiers and then this should be good.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13011940778/job/36302275937)
Build commit: 840317a990493de24cf139e520ebc5459b787864 (`840317a990493de24cf139e520ebc5459b787864`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13011940778/job/36302275937"},"merge_sha":"840317a990493de24cf139e520ebc5459b787864"} -->

jieyouxu commented:
Changes since last review:

- Fixed symbol helper comment.
- Dropped symbol checks on binary.
- Changed `hello` fn names to be different between the various input rlib/dylib crates.

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit ebdcf01b346bf8b5d6699686e2624aed643bfa15 with merge 1395e249256ef3527df6193bb9a86b7f9649ae3f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ebdcf01b346bf8b5d6699686e2624aed643bfa15","merge_sha":"1395e249256ef3527df6193bb9a86b7f9649ae3f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13025258846/job/36338184078)
Build commit: 1395e249256ef3527df6193bb9a86b7f9649ae3f (`1395e249256ef3527df6193bb9a86b7f9649ae3f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13025258846/job/36338184078"},"merge_sha":"1395e249256ef3527df6193bb9a86b7f9649ae3f"} -->

Noratrieb reviewed (approved):
gamer

jieyouxu commented:
Rebased to fix tidy conflict.
@bors r=Noratrieb rollup=iffy (might bounce)

bors commented:
:pushpin: Commit 9734ebb9be2ad760385555e36bb0d065e726d6f5 has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb 9734ebb9be2ad760385555e36bb0d065e726d6f5 -->
<!-- homu: {"type":"Approved","sha":"9734ebb9be2ad760385555e36bb0d065e726d6f5","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 9734ebb9be2ad760385555e36bb0d065e726d6f5 with merge 9709af79014a150df7807671e6250ed940c488eb...
<!-- homu: {"type":"BuildStarted","head_sha":"9734ebb9be2ad760385555e36bb0d065e726d6f5","merge_sha":"9709af79014a150df7807671e6250ed940c488eb"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13082982539/job/36513915475)
Approved by: Noratrieb
Pushing 9709af79014a150df7807671e6250ed940c488eb to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Noratrieb","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13082982539/job/36513915475"},"merge_sha":"9709af79014a150df7807671e6250ed940c488eb"} -->

rust-timer commented:
Finished benchmarking commit (9709af79014a150df7807671e6250ed940c488eb): [comparison URL](https://perf.rust-lang.org/compare.html?start=854f22563c8daf92709fae18ee6aed52953835cd&end=9709af79014a150df7807671e6250ed940c488eb&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=854f22563c8daf92709fae18ee6aed52953835cd&end=9709af79014a150df7807671e6250ed940c488eb&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.4%, 0.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=854f22563c8daf92709fae18ee6aed52953835cd&end=9709af79014a150df7807671e6250ed940c488eb&stat=max-rss)
<details>
<summary>Results (primary 3.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.2% | [3.2%, 3.2%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.2% | [3.2%, 3.2%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=854f22563c8daf92709fae18ee6aed52953835cd&end=9709af79014a150df7807671e6250ed940c488eb&stat=cycles:u)
<details>
<summary>Results (secondary -2.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-3.6%, -2.3%] | 7     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=854f22563c8daf92709fae18ee6aed52953835cd&end=9709af79014a150df7807671e6250ed940c488eb&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 779.222s -> 778.098s (-0.14%)
**Artifact size**: 328.75 MiB -> 328.85 MiB (0.03%)
